### DISCUSSION TITLE
Alternate HDMI Capture MS2130 > MS2109

### DISCUSSION THREAD

**Discussion by strukiii** [tiny-pilot/tinypilot#1593] [Q&A]
Hey, I bought [this](https://www.aliexpress.com/item/1005004990708725.html#nav-review) hdmi capture card which seems to be an upgrade over the MS2109 recommended and is from the same company.  [This](https://github.com/awawa-dev/HyperHDR/discussions/499) discussion gives a good overview, except I guess from a video graphics perspective for which I'm not too knowledgeable.  The main difference for our purposes is that it outputs raw YUYV at 60fps, so we can't bypass cpu compression to MJPEG with ustreamer as is possible with the MS2109.  I assume this will add too much latency to make it good for kvm usage, but maybe there's a hardware encoder that may be used?  

Regardless, I tried it anyway with the default settings, and I'm not seeming to be able to get a video output.  From the logs, ustreamer appears to fallback and handle the differing format / enable encoding.  Not sure if anyone more knowledgeable has tried this one out, I'm very new to tinkering with tinypilot.  

![Screenshot (874)](https://github.com/tiny-pilot/tinypilot/assets/52466430/fb0c153b-96a7-4e48-94c4-4ace5a092ae8)
![Screenshot (875)](https://github.com/tiny-pilot/tinypilot/assets/52466430/7eb126e0-fa3f-4277-a7d3-c66aff55773a)
```Aug 26 18:16:22 raspberrypi systemd[1]: Started uStreamer - Lightweight, optimized video encoder.
Aug 26 18:16:22 raspberrypi launch[943]: + exec /opt/ustreamer/ustreamer --port 48001 --persistent --h264-sink tinypilot::ustreamer::h264 --h264-sink-mode 777 --h264-sink-rm --encoder hw --format jpeg --resolution 1920x1080 --h264-bitrate 20000
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.579      main] -- Starting PiKVM uStreamer 5.38 ...
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.579      main] -- Using internal blank placeholder
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.579      main] -- Using H264-sink: tinypilot::ustreamer::h264
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.580      main] -- Listening HTTP on [127.0.0.1]:48001
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.580    stream] -- Using V4L2 device: /dev/video0
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.580    stream] -- Using desired FPS: 0
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.580      http] -- Starting HTTP eventloop ...
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.580    stream] -- H264: Initializing encoder ...
Aug 26 18:16:22 raspberrypi launch[943]: -- INFO  [4250.588    stream] -- H264: Configuring encoder: DMA=0 ...
Aug 26 18:16:23 raspberrypi launch[943]: ================================================================================
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.605    stream] -- Device fd=10 opened
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.605    stream] -- Using input channel: 0
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.621    stream] -- Using resolution: 1920x1080
Aug 26 18:16:23 raspberrypi launch[943]: -- ERROR [4250.621    stream] -- Could not obtain the requested format=JPEG; driver gave us YUYV
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.621    stream] -- Falling back to format=YUYV
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.621    stream] -- Using format: YUYV
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.638    stream] -- Using HW FPS: 0 -> 60 (coerced)
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.638    stream] -- Using IO method: MMAP
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.652    stream] -- Requested 5 device buffers, got 5
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.657    stream] -- Capturing started
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.657    stream] -- Switching to CPU encoder: the input format is not (M)JPEG ...
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.657    stream] -- Using JPEG quality: 80%
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.657    stream] -- Creating pool JPEG with 4 workers ...
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.657    stream] -- Capturing ...
Aug 26 18:16:23 raspberrypi launch[943]: -- INFO  [4250.686    stream] -- H264: Configuring encoder: DMA=1 ...
Aug 26 18:17:55 raspberrypi launch[943]: -- INFO  [4343.369      http] -- HTTP: NEW client (now=1): [fe80::f9e:4922:6ef1:dac6]:56828, id=984335bad1a246a4```

**Comment by mtlynch** [comment]
Oh, thanks for the pointer! I didn't realize there was a new HDMI to USB capture chip available.

TinyPilot is configured to run with the MS2109 settings by default. You can try experimenting with other settings like this:

```bash
sudo ustreamer stop
sudo /opt/ustreamer/ustreamer \
  --port 48001 \
  --format yuyv \
  --encoder m2m-image \
  --workers 3 \
  --persistent \
  --h264-sink tinypilot::ustreamer::h264 \
  --h264-sink-mode 777 \
  --h264-sink-rm \
  --drop-same-frames 30 \
  --dv-timings
```

In that one, I've changed it to `yuyv` format and set the encoder as `m2m-image`, which will use the Pi's GPU.  The `--encoder hw` flag tells uStreamer to accept MJPEG directly from the hardware. That works on the MS2109, but maybe it doesn't on `MS2130`.

**Reply by strukiii** [reply]
Thanks for the pointer, I'll give this a shot and report back :)

**Comment by wsxws** [comment]
there is no further report back, kann i asume that the setting works for 2130 ?

**Reply by strukiii** [reply]
Hey, I did manage to get it working!  I got very busy and shelved the project though.  I had to slightly tweak the settings from those given above but video did go through ustreamer.  One Issue is the software encoder had to be used, so you likely will see better results with the older card.
